**Zadanie1**: Jaką minimalną i maksymalną wartość może przyjmować współczynnik Giniego? Uzasadnij. Co to wtedy oznacza?

Współczynnik Giniego mieści się w przedziale od 0 do 1. Wartość 0 oznacza, że grupa jest rozdzielona całkowicie, tj. mamy punkt czysty. Poprzez wartość 1 rozumie się, że elementy są bardzo zróżnicowane i należą do różnych hipotetycznych grup. W toerii dążymy do osiągnięcia punktów czystych, czyli zmniejszenia współczynnika Giniego.

**Zadanie2**: Uzasadnij wartość entropii w wierzchołku (u samej góry) poprzez bezpośrednie obliczenia.

In [ ]:
import math
r1= 34/105
r2= 32/105
r3= 39/105
log2r1= math.log2(r1)
log2r2= math.log2(r2)
log2r3= math.log2(r3)
entropia= (-r1*log2r1)+(-r2*log2r2)+(-r3*log2r3)
entropia

1.5799176782552726

**Zadanie3**: Uzasadnij wartość indeksu Giniego w wierzchołku (u samej góry) poprzez bezpośrednie obliczenia.

In [ ]:
gini= 1-r1**2-r2**2-r3**2
gini

0.6643083900226757

**Zadanie4**: Przeczytaj https://predictivesolutions.pl/jak-udoskonalic-algorytm-drzew-decyzyjnych. Sporządź odpowiednią notatkę.

Notatka załączona w pliku pdf :)

**Zadanie5**: Modele lasu drzew decyzyjnych często wykorzystują metodę modyfikowania danych treningowych - agregację bootstrapową (bootstraping). Polega ona na wielokrotnym losowaniu ze zwracaniem. Jeżeli takie losowanie powtózymy  𝑛  razy, otrzymamy  𝑛  elementowy zbiór danych treningowych, w którym część przypadków będzie się powtarzać. Pokaż, że dla dużych  𝑛  próba będzie zawierała średnio  63%  przypadków z orginalnego zbioru.

In [ ]:
x= range(10000)
import random
y= random.choices(x, k=10000)
len(set(y))/10000

0.6347

In [ ]:
n=1000
1-(1-1/n)**n

0.6323045752290363

Miałam odpowiedzieć na pytanie dlaczego tak wychodzi z tego wzoru. Więc najpierw zajęłam się (1-1/n)**n. Jako, że jest to liczba jest operacja na liczbach n-tych obliczyłam granicę z n dążacego do nieskończoności. Wyszło mi 1/e. Pod e podłożyłam 2,72. Z końcowego obliczenia wychodziło 1,72/2,72=0.632 :)

Dodatkowo ważne jest max_features, które ustala tą losowość. Jak wysoka to oznacza, że drzewa będą do siebie podobne i będą łatwo dopasowywać się do danych. Niska wartość z kolei oznacza, że drzewa będą zupełnie inne i że każde będzie musiało być dość głębokie, aby dobrze się dopasować do danych.

**Zadanie6**: Rozważ dane zawierające nagłówki, przy czym są to fake newsy (1298) oraz prawdziwe tytuły (1968). Jeden nagłówek to jeden rekord.
Wyznacz częstości występowania wszystkich słów ze wszystkich nagłówków.

Jakie słowa (oprócz stopwordsów) najczęściej pojawiały się w realnych a jakie w fałszywych nagłówkach?

Będziemy tworzyć klasyfiaktor dla tych danych w oparciu o regresję logistyczną oraz drzewa losowe/lasy losowe. Każdy nagłówek będzie reprezentowany w postaci wektora zer i jedynek w zależności od występowania danego słowa (długość wektora = liczba wszystkich unikatowych słów, może warto jednak zawęzić? albo potraktować jako hiperparametr). Podziel dane na 3 grupy: 70% zbiór treningowy, 15% zbiór walidacyjny, 15% zbiór testowy. Przetestuj różne zestawy hiperparametrów na zbiorze walidacyjnym. Skomentuj otrzymane wyniki.

¶

In [21]:
#działam na pliku real
from sklearn.feature_extraction.text import CountVectorizer
vect= CountVectorizer()
real=['real.txt']
vect.fit(real) #tokenizacja
bag_of_words= vect.transform(real) #tworzenie worka słów

def data_split_group(data): #dzielę dane na trzy grupy
  train = data[:round(0.7*len(data))]
  test = data[round(0.7*len(data)):round(0.15*len(data))]
  val = data[round(0.15*len(data)):]
data_split_group(real)

vect= CountVectorizer().fit(real) #przetwarzam dane
X_train= vect.transform(real)
print("X_train:\n{}".format(repr(X_train))) #hmmm coś za mała ta macierz...

#stop words
from sklearn.feature_extraction.text import ENGLISH_STOP_WORDS
vect2= CountVectorizer(min_df=0.5, stop_words="english").fit(real)
vect.transform(real) #tu nie mogę nic sensownego uzyskać, ponieważ musi być błąd na etapie tokenizacji



X_train:
<1x2 sparse matrix of type '<class 'numpy.int64'>'
	with 2 stored elements in Compressed Sparse Row format>


ValueError: ignored